src/Tmds.Ssh/KerberosCredential.cs

```diff
@@ -0,0 +1,339 @@
+        {
+            // Remove once .NET 8 is no longer the minimum and we get rid of reflection.
+            return false;
+        }
+
+        return true;
+    }
+
+    private static async Task<bool> TryStageOid(SshConnection connection, ILogger logger, string userName, CancellationToken ct)
+    {
+        {
+            using var userAuthMsg = CreateOidRequestMessage(connection.SequencePool,
+                                        userName, KRB5_OID);
+            await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+        }
+
+        while (true)
+        {
+            using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+            MessageId? mid = response.MessageId;
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_BANNER)
+            {
+                // First message in exchange could be a banner which we ignore
+                // for now.
+                continue;
+            }
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_FAILURE)
+            {
+                logger.AuthenticationKerberosFailed("No OID response received");
+                return false;
+            }
+            else if (mid != MessageId.SSH_MSG_USERAUTH_GSSAPI_RESPONSE)
+            {
+                ThrowHelper.ThrowProtocolUnexpectedValue();
+            }
+
+            ReadOnlySequence<byte> oidResponse = GetGssapiOidResponse(response);
+            if (KRB5_OID.AsSpan().SequenceEqual(oidResponse.IsSingleSegment ? oidResponse.FirstSpan : oidResponse.ToArray()))
+            {
+                return true;
+            }
+            else
+            {
+                var reader = new AsnReader(oidResponse.ToArray(), AsnEncodingRules.DER);
+                string receivedOid = reader.ReadObjectIdentifier();
+                logger.AuthenticationKerberosFailed($"OID response {receivedOid} did not match expected value");
+                return false;
+            }
+        }
+    }
+
+    private static async Task<bool> TryStageAuthentication(SshConnection connection, ILogger logger, NegotiateAuthentication authContext, CancellationToken ct)
+    {
+        byte[]? outToken;
+        NegotiateAuthenticationStatusCode statusCode;
+        try
+        {
+            outToken = authContext.GetOutgoingBlob(Array.Empty<byte>(), out statusCode);
+        }
+        catch (InvalidOperationException)
+        {
+            logger.AuthenticationKerberosFailed("Kerberos library not available");
+            return false;
+        }
+
+        while (outToken is not null)
+        {
+            {
+                using var userAuthMsg = CreateGssapiTokenMessage(connection.SequencePool, outToken);
+                await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+            }
+
+            // If the context is complete we don't expect a response.
+            if (statusCode == NegotiateAuthenticationStatusCode.Completed)
```

Comment:
Can we change this condition to one that checks the status for continuing the loop?
Perhaps: `if (statusCode != NegotiateAuthenticationStatusCode.ContinueNeeded`)?

Reply:
Sure I'll have a look, I think this new logic should work.